provider/core/src/main/java/org/arquillian/pact/provider/core/InteractionRunner.java

```diff
@@ -201,23 +217,30 @@ private Field validateAndGetResourceField(TestClass testClass, Class<?> fieldTyp
     }
 
     protected void executeStateChanges(final RequestResponseInteraction interaction, final TestClass testClass, final Object target) {
-        if (interaction.getProviderState() != null && !interaction.getProviderState().isEmpty()) {
-            final String state = interaction.getProviderState();
-            for (Method ann : testClass.getMethods(State.class)) {
-                if (ArrayUtils.contains(ann.getAnnotation(State.class).value(), state)) {
-                    try {
-                        ann.invoke(target);
-                    } catch (IllegalAccessException e) {
-                        throw new IllegalArgumentException(e);
-                    } catch (InvocationTargetException e) {
-                        throw new IllegalArgumentException(e);
-                    }
-                }
+        if (!interaction.getProviderStates().isEmpty()) {
+            for (final ProviderState state : interaction.getProviderStates()) {
+                Arrays.stream(testClass.getMethods(State.class))
```

Comment:
Don't know if it is a good idea (in terms of readability) to add an stream inside an stream, this will be translated to:

```
Arrays.stream(i.getProviderStates()).foreach(state -> {
     Arrays.stream(state.getMethods())
    .filter(....)
)